File: c/waitset/src/subscriber.c
/*
 * c/waitset -- subscriber. Direct C port of zig/waitset/subscriber.zig; see
 * docs/design/waitset-reference-app.md at the repo root for the full spec.
 *
 * One WaitSet, four conditions attached simultaneously:
 *
 *   - StatusCondition (SUBSCRIPTION_MATCHED_STATUS) -- logs the match.
 *   - QueryCondition ("priority > %0", param "4") -- real attach/trigger/
 *     query-expression/parameters exercise. Drained via the generated
 *     WaitsetSampleDataReader_take_w_condition (what the OMG spec calls
 *     take_w_condition, and part of the typed DataReader's implicit IDL on
 *     every binding -- see zidl's roadmap for the fuller writeup on closing
 *     this gap).
 *   - ReadCondition (any sample/view/instance state) -- same trigger
 *     condition as QueryCondition above.
 *   - GuardCondition -- same watchdog-thread pattern as the publisher.
 *
 * Like publisher.c, branches on membership in wait()'s returned
 * DDS_ConditionSeq -- see that file's comment for the full reasoning.
 *
 * Required stdout markers: "Create topic:", "Create reader for topic:",
 * "Subscriber: writer matched", "Subscriber: high-priority count=",
 * "Subscriber: low-priority count=", "Subscriber: received all 10
 * samples.", "Subscriber: done." Any failure path prints a line starting
 * "FAIL:" and exits nonzero.
 */
#include "waitset_sample.h"
#include "zzdds_c.h"
#include "zzdds.h"

#include <pthread.h>
#include <stdatomic.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define EXPECTED_SAMPLES 10
#define HIGH_PRIORITY_THRESHOLD "4" /* priority > 4 => count 5..9 */
#define WAIT_STEP_SEC 1
#define OVERALL_DEADLINE_MS 30000
#define WATCHDOG_POLL_MS 50

typedef struct {
    DDS_GuardCondition gc;
    atomic_bool stop;
} Watchdog;

static void *watchdog_run(void *arg) {
    Watchdog *w = (Watchdog *)arg;
    int elapsed_ms = 0;
    while (!atomic_load(&w->stop)) {
        if (elapsed_ms >= OVERALL_DEADLINE_MS) {
            printf("Watchdog: overall deadline exceeded, triggering GuardCondition\n");
            DDS_GuardCondition_set_trigger_value(w->gc, true);
            return NULL;
        }
        usleep(WATCHDOG_POLL_MS * 1000);
        elapsed_ms += WATCHDOG_POLL_MS;
    }
    return NULL;
}

static bool condition_active(const DDS_ConditionSeq *active, DDS_Condition c) {
    for (uint32_t i = 0; i < active->_length; i++) {
        if (active->_buffer[i] == c) return true;
    }
    return false;
}

static uint32_t parse_domain(int argc, char **argv) {
    for (int i = 1; i < argc - 1; i++) {
        if (strcmp(argv[i], "-d") == 0 || strcmp(argv[i], "--domain") == 0) {
            return (uint32_t)strtoul(argv[i + 1], NULL, 10);
        }
    }
    return 0;
}

int main(int argc, char **argv) {
    uint32_t domain_id = parse_domain(argc, argv);

    zzdds_DomainParticipantFactory factory = zzdds_create_factory();
    if (zzdds_factory_is_nil(factory)) {
        fprintf(stderr, "FAIL: createFactory() failed\n");
        return 1;
    }
    DDS_DomainParticipantFactory dds_factory = zzdds_DomainParticipantFactory_as_DDS_DomainParticipantFactory(factory);

    DDS_DomainParticipant dp = DDS_DomainParticipantFactory_create_participant(dds_factory, domain_id, NULL, NULL, 0);
    if (!dp) {
        fprintf(stderr, "FAIL: create_participant() failed on domain %u\n", domain_id);
        return 1;
    }

    /* Unlike hello_world, this registration's get_field callback actually
     * matters here -- the QueryCondition's "priority > %0" expression needs
     * field-level access to evaluate. */
    if (zzdds_register_type_support(dp, "WaitsetSample", WaitsetSample_compute_key_hash_from_cdr, WaitsetSample_get_field_from_cdr) != DDS_RETCODE_OK) {
        fprintf(stderr, "FAIL: register_type_support failed\n");
        return 1;
    }

    DDS_Topic topic = DDS_DomainParticipant_create_topic(dp, "WaitsetSample", "WaitsetSample", NULL, NULL, 0);
    if (!topic) {
        fprintf(stderr, "FAIL: create_topic() failed\n");
        return 1;
    }
    printf("Create topic: WaitsetSample\n");

    DDS_Subscriber sub = DDS_DomainParticipant_create_subscriber(dp, NULL, NULL, 0);
    if (!sub) {
        fprintf(stderr, "FAIL: create_subscriber() failed\n");
        return 1;
    }

    DDS_DataReaderQos dr_qos;
    DDS_Subscriber_get_default_datareader_qos(sub, &dr_qos);
    dr_qos.reliability.kind = DDS_ReliabilityQosPolicyKind_RELIABLE_RELIABILITY_QOS;
    dr_qos.history.kind = DDS_HistoryQosPolicyKind_KEEP_ALL_HISTORY_QOS;

    DDS_TopicDescription topic_desc = zzdds_topic_as_description(topic);
    DDS_DataReader dr = DDS_Subscriber_create_datareader(sub, topic_desc, &dr_qos, NULL, 0);
    if (!dr) {
        fprintf(stderr, "FAIL: create_datareader() failed\n");
        return 1;
    }
    printf("Create reader for topic: WaitsetSample\n");

    /* ── WaitSet setup: all four condition types on one WaitSet ── */

    DDS_WaitSet ws = zzdds_create_waitset();
    if (zzdds_waitset_is_nil(ws)) {
        fprintf(stderr, "FAIL: create_waitset() failed\n");
        return 1;
    }

    DDS_Entity dr_entity = DDS_DataReader_as_DDS_Entity(dr);
    DDS_StatusCondition sc = DDS_Entity_get_statuscondition(dr_entity);
    if (DDS_StatusCondition_set_enabled_statuses(sc, DDS_SUBSCRIPTION_MATCHED_STATUS) != DDS_RETCODE_OK) {
        fprintf(stderr, "FAIL: set_enabled_statuses() failed\n");
        return 1;
    }

    DDS_ReadCondition rc_cond = DDS_DataReader_create_readcondition(dr, DDS_ANY_SAMPLE_STATE, DDS_ANY_VIEW_STATE, DDS_ANY_INSTANCE_STATE);
    if (!rc_cond) {
        fprintf(stderr, "FAIL: create_readcondition() failed\n");
        return 1;
    }

    char *param_bufs[1] = {HIGH_PRIORITY_THRESHOLD};
    DDS_StringSeq params;
    params._maximum = 1;
    params._length = 1;
    params._buffer = param_bufs;
    params._release = false;
    DDS_QueryCondition qc_cond = DDS_DataReader_create_querycondition(
        dr, DDS_ANY_SAMPLE_STATE, DDS_ANY_VIEW_STATE, DDS_ANY_INSTANCE_STATE,
        "priority > %0", &params);
    if (!qc_cond) {
        fprintf(stderr, "FAIL: create_querycondition() failed\n");
        return 1;
    }

    DDS_GuardCondition gc = zzdds_create_guardcondition();
    if (zzdds_guardcondition_is_nil(gc)) {
        fprintf(stderr, "FAIL: create_guardcondition() failed\n");
        return 1;
    }

    /* Computed once and reused for both attach and the wait loop's
     * membership checks below -- each is `==`-comparable against whatever
     * wait() later returns for the same underlying condition, including
     * qc_cond's two-level QueryCondition -> ReadCondition -> Condition
     * upcast. */
    DDS_Condition sc_cond = DDS_StatusCondition_as_DDS_Condition(sc);
    DDS_Condition rc_as_cond = DDS_ReadCondition_as_DDS_Condition(rc_cond);
    DDS_Condition qc_as_cond = DDS_ReadCondition_as_DDS_Condition(DDS_QueryCondition_as_DDS_ReadCondition(qc_cond));
    DDS_Condition gc_cond = DDS_GuardCondition_as_DDS_Condition(gc);

    if (DDS_WaitSet_attach_condition(ws, sc_cond) != DDS_RETCODE_OK ||
        DDS_WaitSet_attach_condition(ws, rc_as_cond) != DDS_RETCODE_OK ||
        DDS_WaitSet_attach_condition(ws, qc_as_cond) != DDS_RETCODE_OK ||
        DDS_WaitSet_attach_condition(ws, gc_cond) != DDS_RETCODE_OK)
    {
        fprintf(stderr, "FAIL: attach_condition() failed\n");
        return 1;
    }

    Watchdog watchdog;
    watchdog.gc = gc;
    atomic_init(&watchdog.stop, false);
    pthread_t watchdog_thread;
    pthread_create(&watchdog_thread, NULL, watchdog_run, &watchdog);

    /* ── Main loop: wait, branch on which conditions triggered ── */

    WaitsetSampleDataReader reader;
    WaitsetSampleDataReader_init(&reader, dr);
    DDS_Duration_t wait_step = {WAIT_STEP_SEC, 0};
    int received = 0;
    bool matched_logged = false;
    while (received < EXPECTED_SAMPLES) {
        DDS_ConditionSeq active;
        memset(&active, 0, sizeof(active));
        DDS_ReturnCode_t wr = DDS_WaitSet_wait(ws, &active, &wait_step);
        if (wr == DDS_RETCODE_TIMEOUT) continue;
        if (wr != DDS_RETCODE_OK) {
            fprintf(stderr, "FAIL: WaitSet.wait() returned %d\n", wr);
            return 1;
        }

        bool gc_triggered = condition_active(&active, gc_cond);
        bool sc_triggered = condition_active(&active, sc_cond);
        bool query_triggered = condition_active(&active, qc_as_cond);
        bool read_triggered = condition_active(&active, rc_as_cond);
        DDS_ConditionSeq_free(&active);

        if (gc_triggered) {
            fprintf(stderr, "FAIL: watchdog fired -- only received %d/%d samples\n", received, EXPECTED_SAMPLES);
            return 1;
        }
        if (!matched_logged && sc_triggered) {
            DDS_SubscriptionMatchedStatus status;
            DDS_DataReader_get_subscription_matched_status(dr, &status);
            if (status.current_count > 0) {
                printf("Subscriber: writer matched\n");
                matched_logged = true;
            }
        }

        if (!query_triggered && !read_triggered) continue;

        /* Drain the high-priority subset first via take_w_condition, then
         * whatever's left via a plain take_n. These are two separate calls,
         * each independently locking/unlocking the reader -- a sample can
         * arrive (from the RTPS receive thread) in the gap between them,
         * missing the first (query) take and getting swept up by the
         * second (unfiltered) one. Confirmed as a real, reproducible race
         * in zig/waitset and cpp/waitset (not just in theory): a
         * high-priority sample occasionally printed as "low-priority"
         * (still correctly *received*, just mislabeled) despite
         * take_w_condition's own filtering being correct at the instant it
         * ran. So: still call both -- take_w_condition is still genuinely
         * exercised, and still does the real draining -- but decide the
         * *label* from each sample's own already-deserialized `priority`
         * field rather than trusting which call it came from. */
        WaitsetSample high_values[EXPECTED_SAMPLES];
        zzdds_sample_info high_infos[EXPECTED_SAMPLES];
        memset(high_values, 0, sizeof(high_values));
        memset(high_infos, 0, sizeof(high_infos));
        int n_high = WaitsetSampleDataReader_take_w_condition(
            &reader, DDS_QueryCondition_as_DDS_ReadCondition(qc_cond), high_values, high_infos, EXPECTED_SAMPLES);

        WaitsetSample low_values[EXPECTED_SAMPLES];
        zzdds_sample_info low_infos[EXPECTED_SAMPLES];
        memset(low_values, 0, sizeof(low_values));
        memset(low_infos, 0, sizeof(low_infos));
        int n_low = WaitsetSampleDataReader_take_n(&reader, low_values, low_infos, EXPECTED_SAMPLES,
                                                     DDS_ANY_SAMPLE_STATE, DDS_ANY_VIEW_STATE, DDS_ANY_INSTANCE_STATE);

        for (int i = 0; i < n_high; i++) {
            if (!high_infos[i].valid_data) continue;
            if (high_values[i].priority > 4) {
                printf("Subscriber: high-priority count=%d priority=%d\n", high_values[i].count, high_values[i].priority);
            } else {
                printf("Subscriber: low-priority count=%d priority=%d\n", high_values[i].count, high_values[i].priority);
            }
            received++;
        }
        for (int i = 0; i < n_low; i++) {
            if (!low_infos[i].valid_data) continue;
            if (low_values[i].priority > 4) {
                printf("Subscriber: high-priority count=%d priority=%d\n", low_values[i].count, low_values[i].priority);
            } else {
                printf("Subscriber: low-priority count=%d priority=%d\n", low_values[i].count, low_values[i].priority);
            }
            received++;
        }
    }

    printf("Subscriber: received all %d samples.\n", EXPECTED_SAMPLES);

    atomic_store(&watchdog.stop, true);
    pthread_join(watchdog_thread, NULL);

    /* Well-behaved cleanup: detach and delete every condition explicitly
     * before tearing the reader down (contrast with publisher.c, which
     * deliberately does *not* do this for its StatusCondition). */
    DDS_WaitSet_detach_condition(ws, DDS_StatusCondition_as_DDS_Condition(sc));
    DDS_WaitSet_detach_condition(ws, DDS_ReadCondition_as_DDS_Condition(rc_cond));
    DDS_WaitSet_detach_condition(ws, DDS_ReadCondition_as_DDS_Condition(DDS_QueryCondition_as_DDS_ReadCondition(qc_cond)));
    DDS_WaitSet_detach_condition(ws, DDS_GuardCondition_as_DDS_Condition(gc));
    DDS_DataReader_delete_readcondition(dr, rc_cond);
    DDS_DataReader_delete_readcondition(dr, DDS_QueryCondition_as_DDS_ReadCondition(qc_cond));
    DDS_Subscriber_delete_datareader(sub, dr);

    zzdds_destroy_guardcondition(gc);
    zzdds_destroy_waitset(ws);

    printf("Subscriber: done.\n");
    zzdds_destroy_factory(factory);
    return 0;
}

File: java/waitset/Subscriber.java
// java/waitset -- subscriber. Direct Java port of
// zig/waitset/subscriber.zig / c/waitset/src/subscriber.c /
// cpp/waitset/src/subscriber.cpp; see docs/design/waitset-reference-app.md
// at the repo root for the full spec.
//
// One WaitSet, four conditions attached simultaneously:
//
//   - StatusCondition (SUBSCRIPTION_MATCHED_STATUS) -- logs the match.
//   - QueryCondition ("priority > %0", param "4") -- real attach/trigger/
//     query-expression/parameters exercise. Drained via the generated
//     WaitsetSampleDataReader.take_w_condition (what the OMG spec calls
//     take_w_condition, and part of the typed DataReader's implicit IDL on
//     every binding -- see zidl's roadmap for the fuller writeup on closing
//     this gap). QueryCondition satisfies ReadCondition directly via Java's
//     own interface inheritance -- no upcast needed, unlike C/C++'s
//     as_ReadCondition()/native_handle() dance.
//   - ReadCondition (any sample/view/instance state) -- same trigger
//     condition as QueryCondition above.
//   - GuardCondition -- same watchdog-thread pattern as the publisher.
//
// Like publisher.java, branches on each held condition's own
// get_trigger_value() directly rather than membership in wait()'s returned
// active_conditions list.
//
// Required stdout markers: "Create topic:", "Create reader for topic:",
// "Subscriber: writer matched", "Subscriber: high-priority count=",
// "Subscriber: low-priority count=", "Subscriber: received all 10
// samples.", "Subscriber: done." Any failure path prints a line starting
// "FAIL:" and exits nonzero.

import io.zzdds.dcps.Dcps;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.concurrent.atomic.AtomicBoolean;

public class Subscriber {
    static final int EXPECTED_SAMPLES = 10;
    static final String HIGH_PRIORITY_THRESHOLD = "4"; // priority > 4 => count 5..9
    static final int WAIT_STEP_SEC = 1;
    static final int OVERALL_DEADLINE_MS = 30000;
    static final int WATCHDOG_POLL_MS = 50;

    static int parseDomain(String[] args) {
        for (int i = 0; i < args.length - 1; i++) {
            if (args[i].equals("-d") || args[i].equals("--domain")) {
                return Integer.parseInt(args[i + 1]);
            }
        }
        return 0;
    }

    static class Watchdog implements Runnable {
        final Dcps.DDS.GuardCondition gc;
        final AtomicBoolean stop = new AtomicBoolean(false);

        Watchdog(Dcps.DDS.GuardCondition gc) { this.gc = gc; }

        public void run() {
            int elapsedMs = 0;
            while (!stop.get()) {
                if (elapsedMs >= OVERALL_DEADLINE_MS) {
                    System.out.println("Watchdog: overall deadline exceeded, triggering GuardCondition");
                    gc.set_trigger_value(true);
                    return;
                }
                try {
                    Thread.sleep(WATCHDOG_POLL_MS);
                } catch (InterruptedException e) {
                    return;
                }
                elapsedMs += WATCHDOG_POLL_MS;
            }
        }
    }

    public static void main(String[] args) throws Exception {
        int domainId = parseDomain(args);

        Dcps.DDS.DomainParticipantFactory factory =
            (Dcps.DDS.DomainParticipantFactory) io.zzdds.runtime.ZzddsRuntime.createFactory();
        if (factory == null) {
            System.err.println("FAIL: createFactory() failed");
            System.exit(1);
        }

        Dcps.DDS.DomainParticipant dp = factory.create_participant(domainId, null, null, 0);
        if (dp == null) {
            System.err.println("FAIL: create_participant() failed on domain " + domainId);
            System.exit(1);
        }

        // Unlike hello_world, this registration's get_field callback
        // actually matters here -- the QueryCondition's "priority > %0"
        // expression needs field-level access to evaluate.
        if (WaitsetSampleTypeSupport.register(dp, "WaitsetSample") != 0) {
            System.err.println("FAIL: register_type_support failed");
            System.exit(1);
        }

        Dcps.DDS.Topic topic = dp.create_topic("WaitsetSample", "WaitsetSample", null, null, 0);
        if (topic == null) {
            System.err.println("FAIL: create_topic() failed");
            System.exit(1);
        }
        System.out.println("Create topic: WaitsetSample");

        Dcps.DDS.Subscriber sub = dp.create_subscriber(null, null, 0);
        if (sub == null) {
            System.err.println("FAIL: create_subscriber() failed");
            System.exit(1);
        }

        Dcps.DDS.DataReaderQos drQos = new Dcps.DDS.DataReaderQos();
        sub.get_default_datareader_qos(drQos);
        drQos.get_reliability().set_kind(Dcps.DDS.ReliabilityQosPolicyKind.RELIABLE_RELIABILITY_QOS);
        drQos.get_history().set_kind(Dcps.DDS.HistoryQosPolicyKind.KEEP_ALL_HISTORY_QOS);

        // Topic extends TopicDescription directly in Java's interface
        // hierarchy (unlike C++, which needs an explicit upcast) -- no
        // conversion needed.
        Dcps.DDS.DataReader dr = sub.create_datareader(topic, drQos, null, 0);
        if (dr == null) {
            System.err.println("FAIL: create_datareader() failed");
            System.exit(1);
        }
        System.out.println("Create reader for topic: WaitsetSample");

        // ── WaitSet setup: all four condition types on one WaitSet ──

        Dcps.DDS.WaitSet ws = (Dcps.DDS.WaitSet) io.zzdds.runtime.ZzddsRuntime.createWaitSet();
        if (ws == null) {
            System.err.println("FAIL: createWaitSet() failed");
            System.exit(1);
        }

        Dcps.DDS.StatusCondition sc = dr.get_statuscondition();
        if (sc.set_enabled_statuses(Dcps.DDS.SUBSCRIPTION_MATCHED_STATUS.value) != 0) {
            System.err.println("FAIL: set_enabled_statuses() failed");
            System.exit(1);
        }

        Dcps.DDS.ReadCondition rcCond = dr.create_readcondition(
            Dcps.DDS.ANY_SAMPLE_STATE.value, Dcps.DDS.ANY_VIEW_STATE.value, Dcps.DDS.ANY_INSTANCE_STATE.value);
        if (rcCond == null) {
            System.err.println("FAIL: create_readcondition() failed");
            System.exit(1);
        }

        Dcps.DDS.QueryCondition qcCond = dr.create_querycondition(
            Dcps.DDS.ANY_SAMPLE_STATE.value, Dcps.DDS.ANY_VIEW_STATE.value, Dcps.DDS.ANY_INSTANCE_STATE.value,
            "priority > %0", Arrays.asList(HIGH_PRIORITY_THRESHOLD));
        if (qcCond == null) {
            System.err.println("FAIL: create_querycondition() failed");
            System.exit(1);
        }

        Dcps.DDS.GuardCondition gc = (Dcps.DDS.GuardCondition) io.zzdds.runtime.ZzddsRuntime.createGuardCondition();
        if (gc == null) {
            System.err.println("FAIL: createGuardCondition() failed");
            System.exit(1);
        }

        if (ws.attach_condition(sc) != 0 ||
            ws.attach_condition(rcCond) != 0 ||
            ws.attach_condition(qcCond) != 0 ||
            ws.attach_condition(gc) != 0)
        {
            System.err.println("FAIL: attach_condition() failed");
            System.exit(1);
        }

        Watchdog watchdog = new Watchdog(gc);
        Thread watchdogThread = new Thread(watchdog);
        watchdogThread.start();

        // ── Main loop: wait, branch on which conditions triggered ──

        WaitsetSampleDataReader reader = new WaitsetSampleDataReader(dr);
        Dcps.DDS.Duration_t waitStep = new Dcps.DDS.Duration_t();
        waitStep.set_sec(WAIT_STEP_SEC);
        waitStep.set_nanosec(0);

        int received = 0;
        boolean matchedLogged = false;
        while (received < EXPECTED_SAMPLES) {
            ArrayList<Dcps.DDS.Condition> active = new ArrayList<>();
            int wr = ws.wait(active, waitStep);
            if (wr == Dcps.DDS.RETCODE_TIMEOUT.value) continue;
            if (wr != 0) {
                System.err.println("FAIL: WaitSet.wait() returned " + wr);
                System.exit(1);
            }
            if (gc.get_trigger_value()) {
                System.err.println("FAIL: watchdog fired -- only received " + received + "/" + EXPECTED_SAMPLES + " samples");
                System.exit(1);
            }
            if (!matchedLogged && sc.get_trigger_value()) {
                Dcps.DDS.SubscriptionMatchedStatus status = new Dcps.DDS.SubscriptionMatchedStatus();
                dr.get_subscription_matched_status(status);
                if (status.get_current_count() > 0) {
                    System.out.println("Subscriber: writer matched");
                    matchedLogged = true;
                }
            }

            boolean queryTriggered = qcCond.get_trigger_value();
            boolean readTriggered = rcCond.get_trigger_value();
            if (!queryTriggered && !readTriggered) continue;

            // Drain the high-priority subset first via take_w_condition,
            // then whatever's left via a plain take_n. These are two
            // separate calls, each independently locking/unlocking the
            // reader -- a sample can arrive (from the RTPS receive thread)
            // in the gap between them, missing the first (query) take and
            // getting swept up by the second (unfiltered) one. Confirmed as
            // a real, reproducible race in zig/waitset (not just in
            // theory): a high-priority sample occasionally printed as
            // "low-priority" (still correctly *received*, just mislabeled)
            // despite take_w_condition's own filtering being correct at the
            // instant it ran. So: still call both -- take_w_condition is
            // still genuinely exercised, and still does the real draining
            // -- but decide the *label* from each sample's own
            // already-deserialized `priority` field rather than trusting
            // which call it came from.
            WaitsetSampleDataReader.Sample[] highSamples = reader.take_w_condition(qcCond, EXPECTED_SAMPLES);
            WaitsetSampleDataReader.Sample[] lowSamples = reader.take_n(
                EXPECTED_SAMPLES, Dcps.DDS.ANY_SAMPLE_STATE.value, Dcps.DDS.ANY_VIEW_STATE.value, Dcps.DDS.ANY_INSTANCE_STATE.value);
            for (WaitsetSampleDataReader.Sample s : highSamples) {
                if (!s.validData) continue;
                if (s.data.get_priority() > 4) {
                    System.out.println("Subscriber: high-priority count=" + s.data.get_count() + " priority=" + s.data.get_priority());
                } else {
                    System.out.println("Subscriber: low-priority count=" + s.data.get_count() + " priority=" + s.data.get_priority());
                }
                received++;
            }
            for (WaitsetSampleDataReader.Sample s : lowSamples) {
                if (!s.validData) continue;
                if (s.data.get_priority() > 4) {
                    System.out.println("Subscriber: high-priority count=" + s.data.get_count() + " priority=" + s.data.get_priority());
                } else {
                    System.out.println("Subscriber: low-priority count=" + s.data.get_count() + " priority=" + s.data.get_priority());
                }
                received++;
            }
        }

        System.out.println("Subscriber: received all " + EXPECTED_SAMPLES + " samples.");

        watchdog.stop.set(true);
        watchdogThread.join();

        // Well-behaved cleanup: detach and delete every condition explicitly
        // before tearing the reader down (contrast with publisher.java,
        // which deliberately does *not* do this for its StatusCondition).
        ws.detach_condition(sc);
        ws.detach_condition(rcCond);
        ws.detach_condition(qcCond);
        ws.detach_condition(gc);
        dr.delete_readcondition(rcCond);
        dr.delete_readcondition(qcCond);
        sub.delete_datareader(dr);

        io.zzdds.runtime.ZzddsRuntime.destroyGuardCondition(gc);
        io.zzdds.runtime.ZzddsRuntime.destroyWaitSet(ws);

        System.out.println("Subscriber: done.");
    }
}
